### DISCUSSION TITLE
Advice: Thrust or height control - fluently follow of x,y,z waypoints

### DISCUSSION THREAD

**Discussion by frulloa** [bitcraze/discussions#716] [Q&A]
Good morning. Please, is there an example or which command you suggest to use if I want to follow a series of waypoints (x,y,z) in a fluent way (with fluent I mean, not stop on each waypoint)?
Goal: it is to follow a helical trajectory or similar. I mean to follow points (in a fluent way) in x,y,z

Until now, the CF follows in fluent way X and Y waypoints but with fixed Z. Those points are computed every 0.1 seconds, for example, by the equation of a circle using a **for** loop. To do that I am using cf.param.set_value('flightmode.posSet','1') set absolute mode for positioning, then send_setpoint command. But on this command you need thrust value, hence I do not know how to translate height, which is in meters, to thrust value. 

I tried to do it with send_position_setpoint command but does not work well. The shape that makes the drone does not look like a circle.

**Comment by knmcguire** [comment]
You probably need to share a bit of your code (please keep the example a bit minimal and straight to the point).

Maybe not exactly what you are looking for, but this example [uses the highlevel commander](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/autonomous_sequence_high_level.py). The trajectory was pregenerated with a tool given in the comments and uploaded to the memory of the crazyflie, hence making it capable of running it onboard. It is indeed in the horizontal plane, but it is possible to generate a trajectory with z-axis variation.

**Reply by frulloa** [reply]
Hi, thanks for your response and time. I will try to explain better.

I am working with Loco Position System. I have a function whose main part is (take off, landing, logging, etc. are omitted) below. This code use commander.send_setpoint and works fine. But, doesn't allow me to follow z_pos (because it requires thrust). 
Hence, I tried the same idea with the commander.send_position_setpoint (second code below), here follows x_pos, y_pos, and z_pos. Unfortunately, the accuracy is not as good as whit commander.send_setpoint.

```
#first code using commander.send_setpoint
cf=scf.cf   #obtain CF to manage
cf.param.set_value('flightmode.posSet','1') #set absolute mode for positioning
for t in range(100): #run the circle position updates
            position=circle_next_pos(t,r,offset,direction)
            print('ref:({})'.format(t))
            cf.commander.send_setpoint(position[1],position[0],position[3],int(z*1000))
            time.sleep(0.1)
```
```
 #second code using commander.send_position_setpoint 
       for t in range(120): #run the circle position updates
            position=circle_next_pos(t,r,offset,direction)
            print('ref:(x:{},y:{},z:{})'.format(position[0],position[1],position[2]))
            cf.commander.send_position_setpoint(position[0],position[1],position[2],0)
            time.sleep(0.1)
```
The circle_next_pos is a function with circle equations
```
def circle_next_pos(t,r,offset,direction):
    #t: time value, r: radius of circle, offset: position in circle(Degrees), 
    #direction: direction to go around circle
    x_c = 0   
    y_c = 0  
    t = t*direction
    x_pos = x_c+(r*np.cos(2*np.pi*0.02*t+offset))
    y_pos = y_c+(r*np.sin(2*np.pi*0.02*t+offset))
    z_pos=0.015*t+0.5
    print('ref:({},{},{})'.format(x_pos,y_pos,z_pos))
    desired_pos=(x_pos,y_pos,z_pos,0.0)
    return desired_pos
```

My goal is to test more advanced swarm interaction models. Those models give me x_pos, y_pos, z_pos. The CFs should do something similar to the following image. **Do you think my working approach is good? Or what command or approach will best suit my goal?**:
![swarm](https://github.com/bitcraze/discussions/assets/78211872/4f38126d-ad46-4732-8233-0223e4339383)

**Reply by knmcguire** [reply]
I'm surprised that you got it to work with the send_setpoint at all! But that shouldn't be the case ofcourse.

What I expect is happening, is that the send_position_setpoint is a bit too coarse, as it not only tries to match that setpoint, but match that setpoint with velocity zero, so you'll get a bit of a wiggling effect. I don't know which controller you tried? The default (pid) or Mellinger?

Currently we don't have any smooth trajectory planning, and we know that the [crazyswarm](https://imrclab.github.io/crazyswarm2/) project has capability of sending full state setpoint (position and the velocity it goes through that point), which is available in the crazyflie but we don't have that api in the crazyflie python library (see this [issue](https://github.com/bitcraze/crazyflie-lib-python/issues/372)). 

The only quickest smooth trajectory that you can do soonish, is to try out the highlevel commander example by uploading this trajectory onboard of the crazyflie and executing the trajectory following commands from there. You can do that or try to have a go to  implement the full state setpoint from the issue, or go for crazyswarm.

**Reply by frulloa** [reply]
- Thanks for your explication, and suggestions. I attached a video at the end.

- I didn't change any settings about the controller, hence, I think, it's working with the default one. Do you think something would change if I switch to Mellinger?

- Please, one question related to the high-level commander, I reviewed this [example](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/autonomous_sequence_high_level_compressed.py). How can you define the trajectory? I mean, to what corresponds the values defined in the variable trajectory [ln#52](https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/examples/autonomy/autonomous_sequence_high_level_compressed.py#L52) ?


In the following video, you can see how the swarm of CFs rotates around another CF  (using the code that I mentioned in the first reply).

https://github.com/bitcraze/discussions/assets/78211872/df041c65-e0c6-43de-937f-194ec68887d5

**Reply by knmcguire** [reply]
The video looks pretty smooth to me, but perhaps the top drone is a bit unstable but that can also be due to the positioning.

I don't know about the compressed highlevel commander but I recommend you to look through the code https://github.com/bitcraze/crazyflie-lib-python/blob/e6a47634d9d3deca2414cbbedef25ca487ff7415/cflib/crazyflie/mem/trajectory_memory.py#L103

I was actually talking about the non-compressed HL commander here: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/autonomous_sequence_high_level.py,  which generates the trajectory with this tool:  https://github.com/whoenig/uav_trajectories for a tool to generate. It's creating polynomial fits of the position, and the desired velocity that the quadcopter needs to have while going through it.

**Reply by frulloa** [reply]
Thank you
